### PR TITLE
Feature/handle rate limiting

### PR DESCRIPTION
**Pull request by dannidickson**

### CODE REVIEW

File: code/Workable.php [hunk 1]
```diff
@@ -164,6 +164,23 @@ public function callHttpClient(string $url, array $params = [], string $method =
 
             throw $e;
         }
+        
+       // remaining retries within the current interval.
+       $remainingRequests = (int) $response->getHeader('X-Rate-Limit-Remaining')[0];
```
Comment (andrewandante) [inline code/Workable.php]: If this header isn't set for some reason, what happens? My suspicion is an illegal array offset error, can we handle this case gracefully?

File: code/Workable.php [hunk 2]
```diff
@@ -164,6 +164,23 @@ public function callHttpClient(string $url, array $params = [], string $method =
...
+       $sleepInterval = 10;
+
+        // If we are about to hit the rate-limit, we wait for the next limit interval, or wait 10 seconds.
+        if ($remainingRequests === 1) {
```
Comment (andrewandante) [inline code/Workable.php]: If this value is `1`, that means we got a successful response, if I'm understanding this correctly - we have 1 request available to us in the time period. The idea is that this then sleeps to reset the count, the executes the query again?

Would simply returning the response body be an option here?

File: code/Workable.php [hunk 3]
```diff
@@ -164,6 +164,23 @@ public function callHttpClient(string $url, array $params = [], string $method =
...
+            $nextIntervalTimestamp = $response->getHeader('X-Rate-Limit-Reset')[0];
```
Comment (andrewandante) [inline code/Workable.php]: Same question about unset headers here

File: code/Workable.php (hunk 2 above)
Comment (dannidickson) [reply code/Workable.php]: I've changed my approach. Rate limit is caught in the catch block. Not outside of the try/catch. Avoids having to do a check in the catch and outside of it.

File: code/Workable.php [hunk 4]
```diff
@@ -142,47 +138,67 @@ public function getFullJobs($params = [])
...
 
     /**
-     * Wrapper method to configure the RestfulService, make the call, and handle errors
+     * Sends request to Workable API.
+     * Should it exceed the rate limit, this is caught and put to sleep until the next interval. 
+     * The interval duration is provided by Workable via a header. 
+     * When its awaken, it will call itself again, this repeats until its complete.
+     * This returns a json body from the response.
+     * 
+     * Note: See rate limit docs from Workable https://workable.readme.io/docs/rate-limits
```
Comment (andrewandante) [inline code/Workable.php:147]: Love this!

**Review by andrewandante — APPROVED** [review]
Looks good to me - there are some code style things that I would change, but there's no `.editorconfig` or `phpcs.xml` or anything so it flies.

**Comment by andrewandante** [comment]
@scott1702 could you cast an eye and look at getting this merged please? Also would be good to see if we could get this merged into the core repo at some point

**Pull request merged**